### PR TITLE
Topic fix rcl lifecycle test issue

### PR DESCRIPTION
**Pull request by Barry-Xu-2018**
Related issue is https://github.com/ros2/rcl/issues/713

### CODE REVIEW

File: rcl_lifecycle/src/rcl_lifecycle.c [hunk 1]
```diff
@@ -161,17 +161,21 @@ rcl_lifecycle_transition_fini(
 
   rcl_ret_t ret = RCL_RET_OK;
 
-  if (rcl_lifecycle_state_fini(transition->start, allocator) != RCL_RET_OK) {
-    ret = RCL_RET_ERROR;
+  if (transition->start) {
```
Comment (Karsten1987) [inline rcl_lifecycle/src/rcl_lifecycle.c:164]: why is that needed? Calling `fini/deallocate` on a `nullptr` should be totally fine.

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp [hunk 2]
```diff
@@ -326,6 +332,7 @@ TEST(TestRclLifecycle, state_machine) {
   // Node is null
   ret = rcl_lifecycle_state_machine_fini(&state_machine, nullptr, &allocator);
   EXPECT_EQ(ret, RCL_RET_ERROR);
```
Comment (Karsten1987) [inline rcl_lifecycle/test/test_rcl_lifecycle.cpp:334]: I would expect to see `<< rcl_get_error_string().str` here as well.

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp [hunk 3]
```diff
@@ -133,14 +133,19 @@ TEST(TestRclLifecycle, lifecycle_transition) {
 
   ret = rcl_lifecycle_transition_init(
     &transition, expected_id, &expected_label[0], nullptr, nullptr, &allocator);
-  EXPECT_EQ(ret, RCL_RET_OK);
-  rcutils_reset_error();
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
```
Comment (Karsten1987) [inline rcl_lifecycle/test/test_rcl_lifecycle.cpp:136]: I think we still have to reset the error message after this. Basic rule: Every time we consume the error message, we reset it.

Comment (Karsten1987) [reply rcl_lifecycle/test/test_rcl_lifecycle.cpp:136]: same below

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp [hunk 4]
```diff
@@ -133,14 +133,19 @@ TEST(TestRclLifecycle, lifecycle_transition) {
...
     &transition, expected_id, &expected_label[0], start, nullptr, &allocator);
-  EXPECT_EQ(ret, RCL_RET_OK);
-  rcutils_reset_error();
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  ret = rcl_lifecycle_transition_fini(&transition, &allocator);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
 
+  start = reinterpret_cast<rcl_lifecycle_state_t *>(
+    allocator.allocate(sizeof(rcl_lifecycle_state_t), allocator.state));
+  *start = rcl_lifecycle_get_zero_initialized_state();
```
Comment (Karsten1987) [inline rcl_lifecycle/test/test_rcl_lifecycle.cpp:148]: This is changing the semantics of the test though. We expect the `start` state to be `fini`d as well in the call to `rcl_lifecycle_transition_fini`.

File: rcl_lifecycle/src/rcl_lifecycle.c (hunk 1 above)
Comment (Barry-Xu-2018) [reply rcl_lifecycle/src/rcl_lifecycle.c:164]: After checking codes, Yes.

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp (hunk 4 above)
Comment (Barry-Xu-2018) [reply rcl_lifecycle/test/test_rcl_lifecycle.cpp:148]: Below is original codes

https://github.com/ros2/rcl/blob/944068f3d99d6a4b1130fd62e38c69b96f0e759e/rcl_lifecycle/test/test_rcl_lifecycle.cpp#L139-L148

As you know, I add rcl_lifecycle_state_machine_fini() at line 143, which lead that `start` is freed at rcl_lifecycle_state_machine_fini(). In order to keep original logic, I have to reallocate `start` before next test.

Comment (Karsten1987) [reply rcl_lifecycle/test/test_rcl_lifecycle.cpp:148]: you're absolutely right! Thanks for the clarification.

**Comment by Barry-Xu-2018** [comment]
> It would be great if you could tackle the two issues reported in #713 individually. That makes reviewing/merging a tick faster.
> 
> I would recommend having this PR tackle the `rcutils_reset_error` findings and a follow up PR with the memory leaks.

Since `rcutils_reset_error()` should be called as your said, I update title of #713. #713 is for one issue on memory leaks since missing call *_fini().

**Comment by Barry-Xu-2018** [comment]
@Karsten1987  

Please check again.  
Since change is a little big, I use 'force-pushed'.

File: rcl_lifecycle/test/test_rcl_lifecycle.cpp (hunk 2 above)
Comment (Barry-Xu-2018) [reply rcl_lifecycle/test/test_rcl_lifecycle.cpp:338]: ```EXPECT_EQ(ret, RCL_RET_ERROR) <<  rcl_get_error_string().str;```

It means rcl_get_error_string().str is output if ret !=  RCL_RET_ERROR.   
That is, ret is RCL_RET_OK. So I think we need not to add `<<rcl_get_error_string().str`.   
What do you think ?

**Comment by Karsten1987** [comment]
A couple of things:

When running this, I still see the following output:
```
1: [ RUN      ] TestDefaultStateMachine.default_init
1: 
1: >>> [rcutils|error_handling.c:108] rcutils_set_error_state()
1: This error state is being overwritten:
1: 
1:   'invalid allocator, at /home/karsten/ros2_ws/src/ros2/rcl/rcl_lifecycle/src/transition_map.c:92'
1: 
1: with this new error message:
1: 
1:   'can't free transition map, no allocator given
1: , at /home/karsten/ros2_ws/src/ros2/rcl/rcl_lifecycle/src/transition_map.c:57'
1: 
1: rcutils_reset_error() should be called after error handling to avoid this.
1: <<<
1: [       OK ] TestDefaultStateMachine.default_init (4 ms)
```

Is there a chance you could address these as well? Looks like there's another `rcutils_reset_error` missing.

But then further, when run with `valgrind` I see that this test is leaking around 360 bytes on Linux:
```
==5764== LEAK SUMMARY:
==5764==    definitely lost: 360 bytes in 5 blocks
==5764==    indirectly lost: 264 bytes in 3 blocks
==5764==      possibly lost: 0 bytes in 0 blocks
==5764==    still reachable: 26,347 bytes in 79 blocks
==5764==         suppressed: 0 bytes in 0 blocks
```

Granted, looking at the output, I can't really claim that all of them were introduced by this patch. But when running the same test on master I see _only_ 284 bytes lost. That we leak memory is a problem by itself, but I would like you to verify that this patch indeed introduced another memory leak.

**Comment by Barry-Xu-2018** [comment]
> A couple of things:
> 
> When running this, I still see the following output:
> 
> ```
> 1: [ RUN      ] TestDefaultStateMachine.default_init
> 1: 
> 1: >>> [rcutils|error_handling.c:108] rcutils_set_error_state()
> 1: This error state is being overwritten:
> 1: 
> 1:   'invalid allocator, at /home/karsten/ros2_ws/src/ros2/rcl/rcl_lifecycle/src/transition_map.c:92'
> 1: 
> 1: with this new error message:
> 1: 
> 1:   'can't free transition map, no allocator given
> 1: , at /home/karsten/ros2_ws/src/ros2/rcl/rcl_lifecycle/src/transition_map.c:57'
> 1: 
> 1: rcutils_reset_error() should be called after error handling to avoid this.
> 1: <<<
> 1: [       OK ] TestDefaultStateMachine.default_init (4 ms)
> ```
> 
> Is there a chance you could address these as well? Looks like there's another `rcutils_reset_error` missing.

About the cause of this problem, please look at https://github.com/ros2/rcl/issues/713#issuecomment-658043362.   
I provide one solution. Please look at https://github.com/ros2/rcl/issues/713#issuecomment-659255942.  
Do you think this fixing is suitable for this commit ?  

> But then further, when run with `valgrind` I see that this test is leaking around 360 bytes on Linux:
> 
> ```
> ==5764== LEAK SUMMARY:
> ==5764==    definitely lost: 360 bytes in 5 blocks
> ==5764==    indirectly lost: 264 bytes in 3 blocks
> ==5764==      possibly lost: 0 bytes in 0 blocks
> ==5764==    still reachable: 26,347 bytes in 79 blocks
> ==5764==         suppressed: 0 bytes in 0 blocks
> ```
> 
> Granted, looking at the output, I can't really claim that all of them were introduced by this patch. But when running the same test on master I see _only_ 284 bytes lost. That we leak memory is a problem by itself, but I would like you to verify that this patch indeed introduced another memory leak.

Thanks for your information.  
OK. I will investigate it.

**Comment by Barry-Xu-2018** [comment]
@Karsten1987 

In my environment, I execute `valgrind` before and apply this change.   
Besides, I find the result is unstable. So I execute 5 times for each state.   


- Before (commit: 944068f3d99d6a)
    ```shell
    root@81614b33588d:~/ros2_rolling# tail 1.log
    ==1184== LEAK SUMMARY:
    ==1184==    definitely lost: 700 bytes in 14 blocks
    ==1184==    indirectly lost: 846 bytes in 20 blocks
    ==1184==      possibly lost: 133,920 bytes in 124 blocks
    ==1184==    still reachable: 2,567,391 bytes in 1,905 blocks
    ==1184==         suppressed: 0 bytes in 0 blocks
    ==1184== Rerun with --leak-check=full to see details of leaked memory
    ==1184== 
    ==1184== For lists of detected and suppressed errors, rerun with: -s
    ==1184== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
    root@81614b33588d:~/ros2_rolling# tail 2.log
    ==1198== LEAK SUMMARY:
    ==1198==    definitely lost: 628 bytes in 11 blocks
    ==1198==    indirectly lost: 582 bytes in 17 blocks
    ==1198==      possibly lost: 133,920 bytes in 124 blocks
    ==1198==    still reachable: 2,567,304 bytes in 1,905 blocks
    ==1198==         suppressed: 0 bytes in 0 blocks
    ==1198== Rerun with --leak-check=full to see details of leaked memory
    ==1198== 
    ==1198== For lists of detected and suppressed errors, rerun with: -s
    ==1198== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
    root@81614b33588d:~/ros2_rolling# tail 3.log
    ==1211== LEAK SUMMARY:
    ==1211==    definitely lost: 700 bytes in 14 blocks
    ==1211==    indirectly lost: 846 bytes in 20 blocks
    ==1211==      possibly lost: 115,928 bytes in 117 blocks
    ==1211==    still reachable: 2,564,433 bytes in 1,873 blocks
    ==1211==         suppressed: 0 bytes in 0 blocks
    ==1211== Rerun with --leak-check=full to see details of leaked memory
    ==1211== 
    ==1211== For lists of detected and suppressed errors, rerun with: -s
    ==1211== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
    root@81614b33588d:~/ros2_rolling# tail 4.log
    ==1240== LEAK SUMMARY:
    ==1240==    definitely lost: 676 bytes in 13 blocks
    ==1240==    indirectly lost: 758 bytes in 19 blocks
    ==1240==      possibly lost: 137,280 bytes in 129 blocks
    ==1240==    still reachable: 2,567,369 bytes in 1,902 blocks
    ==1240==         suppressed: 0 bytes in 0 blocks
    ==1240== Rerun with --leak-check=full to see details of leaked memory
    ==1240== 
    ==1240== For lists of detected and suppressed errors, rerun with: -s
    ==1240== ERROR SUMMARY: 2 errors from 1 contexts (suppressed: 0 from 0)
    root@81614b33588d:~/ros2_rolling# tail 5.log
    ==1226== LEAK SUMMARY:
    ==1226==    definitely lost: 748 bytes in 16 blocks
    ==1226==    indirectly lost: 1,022 bytes in 22 blocks
    ==1226==      possibly lost: 135,968 bytes in 128 blocks
    ==1226==    still reachable: 2,575,836 bytes in 1,927 blocks
    ==1226==         suppressed: 0 bytes in 0 blocks
    ==1226== Rerun with --leak-check=full to see details of leaked memory
    ==1226== 
    ==1226== For lists of detected and suppressed errors, rerun with: -s
    ==1226== ERROR SUMMARY: 7 errors from 3 contexts (suppressed: 0 from 0)
    ```
    Definitely lost: 628\~748  
    Indirectly lost: 582\~1022

- After (apply this change)
    ```shell
    root@81614b33588d:~/ros2_rolling# tail 1.log 
    ==1617== LEAK SUMMARY:
    ==1617==    definitely lost: 360 bytes in 5 blocks
    ==1617==    indirectly lost: 264 bytes in 3 blocks
    ==1617==      possibly lost: 0 bytes in 0 blocks
    ==1617==    still reachable: 23,073 bytes in 70 blocks
    ==1617==         suppressed: 0 bytes in 0 blocks
    ==1617== Rerun with --leak-check=full to see details of leaked memory
    ==1617== 
    ==1617== For lists of detected and suppressed errors, rerun with: -s
    ==1617== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
    root@81614b33588d:~/ros2_rolling# tail 2.log 
    ==1630== LEAK SUMMARY:
    ==1630==    definitely lost: 360 bytes in 5 blocks
    ==1630==    indirectly lost: 264 bytes in 3 blocks
    ==1630==      possibly lost: 0 bytes in 0 blocks
    ==1630==    still reachable: 23,073 bytes in 70 blocks
    ==1630==         suppressed: 0 bytes in 0 blocks
    ==1630== Rerun with --leak-check=full to see details of leaked memory
    ==1630== 
    ==1630== For lists of detected and suppressed errors, rerun with: -s
    ==1630== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
    root@81614b33588d:~/ros2_rolling# tail 3.log 
    ==1643== LEAK SUMMARY:
    ==1643==    definitely lost: 432 bytes in 8 blocks
    ==1643==    indirectly lost: 528 bytes in 6 blocks
    ==1643==      possibly lost: 0 bytes in 0 blocks
    ==1643==    still reachable: 23,073 bytes in 70 blocks
    ==1643==         suppressed: 0 bytes in 0 blocks
    ==1643== Rerun with --leak-check=full to see details of leaked memory
    ==1643== 
    ==1643== For lists of detected and suppressed errors, rerun with: -s
    ==1643== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
    root@81614b33588d:~/ros2_rolling# tail 4.log 
    ==1657== LEAK SUMMARY:
    ==1657==    definitely lost: 408 bytes in 7 blocks
    ==1657==    indirectly lost: 440 bytes in 5 blocks
    ==1657==      possibly lost: 0 bytes in 0 blocks
    ==1657==    still reachable: 23,073 bytes in 70 blocks
    ==1657==         suppressed: 0 bytes in 0 blocks
    ==1657== Rerun with --leak-check=full to see details of leaked memory
    ==1657== 
    ==1657== For lists of detected and suppressed errors, rerun with: -s
    ==1657== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
    root@81614b33588d:~/ros2_rolling# tail 5.log 
    ==1670== LEAK SUMMARY:
    ==1670==    definitely lost: 384 bytes in 6 blocks
    ==1670==    indirectly lost: 352 bytes in 4 blocks
    ==1670==      possibly lost: 0 bytes in 0 blocks
    ==1670==    still reachable: 23,073 bytes in 70 blocks
    ==1670==         suppressed: 0 bytes in 0 blocks
    ==1670== Rerun with --leak-check=full to see details of leaked memory
    ==1670== 
    ==1670== For lists of detected and suppressed errors, rerun with: -s
    ==1670== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
    ```
    Definitely lost: 360\~432 
    Indirectly lost: 264\~528

Based on above test result, memory leak is decreased obviously.  
I don't know why there is big difference on original checking.  
About the remaining memory leak issue, I think it is better to create another issue to further check.  
What do you think ?

**Comment by Karsten1987** [comment]
@Barry-Xu-2018 thanks for verifying. I think it makes sense what you propose, to open up a separate issue for the memory leaks.

As for the prints (overwriting the rcutils error), I think it makes sense to apply your changes here as well. It's best to address that issue completely here within the same PR.

**Comment by Barry-Xu-2018** [comment]
@Karsten1987 

>  I think it makes sense what you propose, to open up a separate issue for the memory leaks.

Thank you. 

> As for the prints (overwriting the rcutils error), I think it makes sense to apply your changes here as well. It's best to address that issue completely here within the same PR. Also having it in a PR allows me to review it more easily.

OK. I submit new commit `871601e` to this PR.  
Please review it.

File: rcl_lifecycle/src/default_state_machine.c [hunk 5]
```diff
@@ -659,6 +660,9 @@ rcl_lifecycle_init_default_state_machine(
   rcl_ret_t ret = RCL_RET_ERROR;
   // Used for concatenating error messages in the fail: block.
   const char * fail_error_message = "";
+  char * init_error_string = "";
+  char * fini_error_string = "";
```
Comment (Karsten1987) [inline rcl_lifecycle/src/default_state_machine.c:664]: consider renaming this to `init_error_message` and `fini_error_message` to stay consistent with `fail_error_message`. 
Also I would recommend to add a bit of a docstring on top of each of the three error messages to get an idea on what's going on here.

File: rcl_lifecycle/src/default_state_machine.c [hunk 6]
```diff
@@ -695,14 +699,27 @@ rcl_lifecycle_init_default_state_machine(
...
+    rcl_reset_error();
   }
 
   if (rcl_lifecycle_transition_map_fini(&state_machine->transition_map, allocator) != RCL_RET_OK) {
     const char * fini_error = "";
     if (rcl_error_is_set()) {
-      fini_error = rcl_get_error_string().str;
+      fini_error_string = rcutils_strdup(rcl_get_error_string().str, default_allocator);
+      if (fini_error_string != NULL) {
+        fini_error = rcl_get_error_string().str;
```
Comment (Karsten1987) [inline rcl_lifecycle/src/default_state_machine.c:719]: stay consistent with the logic above
```suggestion
        fini_error = fini_error_string;
```

File: rcl_lifecycle/src/default_state_machine.c (hunk 5 above)
Comment (Karsten1987) [reply rcl_lifecycle/src/default_state_machine.c:664]: Also, why do we need 3 strings here?
I would assume 2 should be sufficient. 

* init_error_message: The original cause or error why we land in `fail:`
* fini_error_message: A conglomerate of failures happening within `fail:`.

File: rcl_lifecycle/src/default_state_machine.c [hunk 7]
```diff
@@ -659,6 +660,9 @@ rcl_lifecycle_init_default_state_machine(
   rcl_ret_t ret = RCL_RET_ERROR;
   // Used for concatenating error messages in the fail: block.
   const char * fail_error_message = "";
+  char * init_error_string = "";
+  char * fini_error_string = "";
+  rcl_allocator_t default_allocator;
```
Comment (Karsten1987) [inline rcl_lifecycle/src/default_state_machine.c:665]: Why not initializing it directly? That would say it later on.

```suggestion
  rcl_allocator_t default_allocator = rcl_get_default_allocator();
```

File: rcl_lifecycle/src/default_state_machine.c [hunk 8]
```diff
@@ -695,14 +699,27 @@ rcl_lifecycle_init_default_state_machine(
...
   // If rcl_lifecycle_transition_map_fini() fails, it will clobber the error string here.
   // Concatenate the error strings if that happens
+  default_allocator = rcl_get_default_allocator();
+
   if (rcl_error_is_set()) {
-    fail_error_message = rcl_get_error_string().str;
+    init_error_string = rcutils_strdup(rcl_get_error_string().str, default_allocator);
+    if (init_error_string != NULL) {
+      fail_error_message = init_error_string;
+    } else {
```
Comment (Karsten1987) [inline rcl_lifecycle/src/default_state_machine.c:708]: I don't think this is necessary. If `init_error_string` is null, then so be it. Only set the `fini_error_message` to the `strdup` failure.

File: rcl_lifecycle/src/default_state_machine.c [hunk 9]
```diff
@@ -695,14 +699,27 @@ rcl_lifecycle_init_default_state_machine(
...
+      fail_error_message = "Error while duplicating strings by rcutils_strdup() !";
+    }
+    rcl_reset_error();
   }
 
   if (rcl_lifecycle_transition_map_fini(&state_machine->transition_map, allocator) != RCL_RET_OK) {
     const char * fini_error = "";
     if (rcl_error_is_set()) {
-      fini_error = rcl_get_error_string().str;
+      fini_error_string = rcutils_strdup(rcl_get_error_string().str, default_allocator);
```
Comment (Karsten1987) [inline rcl_lifecycle/src/default_state_machine.c:717]: So I think this should not be a `strdup` but rather a `snprintf` concatenating two error messages.

File: rcl_lifecycle/src/default_state_machine.c (hunk 6 above)
Comment (Barry-Xu-2018) [reply rcl_lifecycle/src/default_state_machine.c:719]: Agree.   
But as above your suggestion, we will only 2 variables.   
So need not to this assignment-action.

File: rcl_lifecycle/src/default_state_machine.c (hunk 7 above)
Comment (Barry-Xu-2018) [reply rcl_lifecycle/src/default_state_machine.c:665]: `default_allocator` is only used in failure case.  
I think it happens seldom.  
So need not call `rcl_get_default_allocator()` each time.  
What do you think ?

File: rcl_lifecycle/src/default_state_machine.c (hunk 9 above)
Comment (Barry-Xu-2018) [reply rcl_lifecycle/src/default_state_machine.c:717]: The below code has concatenating two error messages. So not do this here.  

https://github.com/ros2/rcl/blob/bd6fbb3d45c271f7351c9d0932b0c7072bd57b1c/rcl_lifecycle/src/default_state_machine.c#L708-L711

File: rcl_lifecycle/src/default_state_machine.c (hunk 8 above)
Comment (Barry-Xu-2018) [reply rcl_lifecycle/src/default_state_machine.c:708]: > If init_error_string is null, then so be it. Only set the fini_error_message to the strdup failure.  

Sorry. I don't understand your comments.  
Your comments describe what to do for else code.  

```c++
    if (init_error_string != NULL) {
      fail_error_message = init_error_string;
    } else {
      fail_error_message = "Error while duplicating strings by rcutils_strdup() !";
    }
```

**Comment by Barry-Xu-2018** [comment]
@Karsten1987 

I rebase branch.  
Based on your review comments, please look at commit `e861e2f`.

**Comment by Karsten1987** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11594/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6735/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9476/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11556/)

**Comment by Barry-Xu-2018** [comment]
@Karsten1987 

Thank you.
